File: reference/typescript/src/knowledge.ts
import { z } from 'zod';

export const KnowledgeCategory = z.enum(['fact', 'preference', 'pattern', 'correction']);
export type KnowledgeCategory = z.infer<typeof KnowledgeCategory>;

export const KnowledgeSource = z.object({
  session_id: z.string().min(1),
  agent_id: z.string().optional(),
  timestamp: z.string().datetime(),
});

export const ProvenanceSource = z.object({
  session_id: z.string().min(1),
  agent_id: z.string().optional(),
  timestamp: z.string().datetime(),
  turn_id: z.string().optional(),
});
export type ProvenanceSource = z.infer<typeof ProvenanceSource>;

export const Provenance = z.object({
  sources: z.array(ProvenanceSource).min(1),
  derived: z.boolean().optional(),
});
export type Provenance = z.infer<typeof Provenance>;

export const GovernanceHandling = z.object({
  retrieval: z.enum(['governed', 'ungoverned']).optional(),
  export: z.enum(['governed', 'ungoverned']).optional(),
  stream: z.enum(['governed', 'ungoverned']).optional(),
});
export type GovernanceHandling = z.infer<typeof GovernanceHandling>;

export const Governance = z.object({
  sensitivity_class: z.enum(['public', 'internal', 'confidential', 'restricted']),
  labels: z.array(z.string().min(1)).default([]),
  handling: GovernanceHandling.optional(),
});
export type Governance = z.infer<typeof Governance>;

export const KnowledgeDecay = z.object({
  half_life_days: z.number().positive().nullable().optional(),
  last_confirmed: z.string().datetime().optional(),
});

export const KnowledgeEntry = z.object({
  oamp_version: z.enum(['1.0.0', '1.1.0', '1.2.0', '1.3.0']),
  type: z.literal('knowledge_entry'),
  id: z.string().uuid(),
  user_id: z.string().min(1),
  category: KnowledgeCategory,
  content: z.string().min(1),
  confidence: z.number().min(0).max(1),
  source: KnowledgeSource,
  provenance: Provenance.optional(),
  governance: Governance.optional(),
  decay: KnowledgeDecay.optional(),
  tags: z.array(z.string()).default([]),
  metadata: z.record(z.unknown()).default({}),
});
export type KnowledgeEntry = z.infer<typeof KnowledgeEntry>;

export const KnowledgeStore = z.object({
  oamp_version: z.enum(['1.0.0', '1.1.0', '1.2.0', '1.3.0']),
  type: z.literal('knowledge_store'),
  user_id: z.string().min(1),
  entries: z.array(KnowledgeEntry),
  exported_at: z.string().datetime(),
  agent_id: z.string().optional(),
  metadata: z.record(z.unknown()).optional(),
});
export type KnowledgeStore = z.infer<typeof KnowledgeStore>;

File: reference/rust/src/knowledge.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

pub const OAMP_VERSION: &str = "1.0.0";

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum KnowledgeCategory {
    Fact,
    Preference,
    Pattern,
    Correction,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KnowledgeSource {
    pub session_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub agent_id: Option<String>,
    pub timestamp: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KnowledgeDecay {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub half_life_days: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub last_confirmed: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KnowledgeEntry {
    pub oamp_version: String,
    #[serde(rename = "type")]
    pub entry_type: String,
    pub id: String,
    pub user_id: String,
    pub category: KnowledgeCategory,
    pub content: String,
    pub confidence: f32,
    pub source: KnowledgeSource,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub decay: Option<KnowledgeDecay>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub tags: Vec<String>,
    #[serde(default, skip_serializing_if = "serde_json::Map::is_empty")]
    pub metadata: serde_json::Map<String, serde_json::Value>,
}

impl KnowledgeEntry {
    pub fn new(
        user_id: &str,
        category: KnowledgeCategory,
        content: &str,
        confidence: f32,
        session_id: &str,
    ) -> Self {
        Self {
            oamp_version: OAMP_VERSION.to_string(),
            entry_type: "knowledge_entry".to_string(),
            id: Uuid::new_v4().to_string(),
            user_id: user_id.to_string(),
            category,
            content: content.to_string(),
            confidence,
            source: KnowledgeSource {
                session_id: session_id.to_string(),
                agent_id: None,
                timestamp: Utc::now(),
            },
            decay: None,
            tags: Vec::new(),
            metadata: serde_json::Map::new(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KnowledgeStore {
    pub oamp_version: String,
    #[serde(rename = "type")]
    pub store_type: String,
    pub user_id: String,
    pub entries: Vec<KnowledgeEntry>,
    pub exported_at: DateTime<Utc>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub agent_id: Option<String>,
}

impl KnowledgeStore {
    pub fn new(user_id: &str, entries: Vec<KnowledgeEntry>) -> Self {
        Self {
            oamp_version: OAMP_VERSION.to_string(),
            store_type: "knowledge_store".to_string(),
            user_id: user_id.to_string(),
            entries,
            exported_at: Utc::now(),
            agent_id: None,
        }
    }
}
